## **Tratamento do Conjunto de Dados** 

In [1]:
import pandas as pd

> Carregando o conjunto de dados como um DataFrame Pandas.

In [2]:
df = pd.read_csv('data\Dados-Abertos.csv', delimiter=';', encoding='latin-1')

In [3]:
df.head()

Nome Ente  UF Código Siafi  Código IBGE   Data   Ano      Mês  \
0  Nova Santa Rosa  PR         7979      4117222  42005  2015  janeiro   
1          Humaitá  RS         8695      4309704  42005  2015  janeiro   
2     Nova Venécia  ES         5677      3203908  42005  2015  janeiro   
3         Gravatal  SC         8121      4206207  42005  2015  janeiro   
4  Passo de Torres  SC         5541      4212254  42005  2015  janeiro   

   Tipo Ente                       OB  CNPJ do Favorecido  \
0  Município  135098000012015OB800001      77116663000109   
1  Município  135098000012015OB800002      87613139000199   
2  Município  135098000012015OB800003      27167428000180   
3  Município  135098000012015OB800004      82926569000147   
4  Município  135098000012015OB800005      95782793000154   

                Nome Favorecido        Nome Emenda Transferência Especial  \
0  MUNICIPIO DE NOVA SANTA ROSA  Emenda Individual                    Não   
1          MUNICIPIO DE HUMAITA  Emenda Individual                    Não   
2     MUNICIPIO DE NOVA VENECIA  Emenda Individual                    Não   
3         MUNICIPIO DE GRAVATAL  Emenda Individual                    Não   
4  MUNICIPIO DE PASSO DE TORRES  Emenda Individual                    Não   

  Categoria Econômica Despesa   Valor  
0         DESPESAS DE CAPITAL  243750  
1         DESPESAS DE CAPITAL  144300  
2         DESPESAS DE CAPITAL   97500  
3         DESPESAS DE CAPITAL  195000  
4         DESPESAS DE CAPITAL   97500

In [4]:
df.dtypes

Nome Ente                      object
UF                             object
Código Siafi                   object
Código IBGE                     int64
Data                            int64
Ano                             int64
Mês                            object
Tipo Ente                      object
OB                             object
CNPJ do Favorecido              int64
Nome Favorecido                object
Nome Emenda                    object
Transferência Especial         object
Categoria Econômica Despesa    object
Valor                          object
dtype: object

In [5]:
example = df.head(1).copy()
example

Nome Ente  UF Código Siafi  Código IBGE   Data   Ano      Mês  \
0  Nova Santa Rosa  PR         7979      4117222  42005  2015  janeiro   

   Tipo Ente                       OB  CNPJ do Favorecido  \
0  Município  135098000012015OB800001      77116663000109   

                Nome Favorecido        Nome Emenda Transferência Especial  \
0  MUNICIPIO DE NOVA SANTA ROSA  Emenda Individual                    Não   

  Categoria Econômica Despesa   Valor  
0         DESPESAS DE CAPITAL  243750

> Checando se existem valores acima da casa dos milhões para a coluna 'Valor'.

In [6]:
million_unit = 7
(df['Valor'].apply(len) > million_unit).any()

True

> Checando qual o tamanho máximo da string para 'Valor'.

In [7]:
df['Comprimento'] = df['Valor'].apply(len)
df['Comprimento'].max()

11

> Obtendo valores de exemplo para os tamanhos.

In [8]:
max_length = 11
filtered = df[df['Comprimento'] == max_length]
filtered.head()

Nome Ente  UF Código Siafi  Código IBGE   Data   Ano       Mês  \
50025         Pernambuco  PE           PE           26  43070  2017  dezembro   
50026         Pernambuco  PE           PE           26  43070  2017  dezembro   
50921              Bahia  BA           BA           29  43160  2018     março   
50932  Rio Grande do Sul  RS           RS           43  43160  2018     março   
61972              Goiás  GO           GO           52  43252  2018     junho   

      Tipo Ente                       OB  CNPJ do Favorecido  \
50025    Estado  530013000012017OB801718      10571982000125   
50026    Estado  530013000012017OB801848      10571982000125   
50921    Estado  135098000012018OB800139      13937057000163   
50932    Estado  135098000012018OB800154      93021632000112   
61972    Estado  135098000012018OB800672      21652711000110   

                                         Nome Favorecido        Nome Emenda  \
50025                               ESTADO DE PERNAMBUCO  Emenda de Bancada   
50026                               ESTADO DE PERNAMBUCO  Emenda de Bancada   
50921  SECRETARIA DA AGRICULTURA, PECUARIA, IRRIGACAO...  Emenda de Bancada   
50932  SECRETARIA DE AGRICULTURA, PECUARIA E DESENVOL...  Emenda de Bancada   
61972  SECRETARIA DE ESTADO DE DESENVOLVIMENTO E INOV...  Emenda de Bancada   

      Transferência Especial Categoria Econômica Despesa        Valor  \
50025                    Não         DESPESAS DE CAPITAL  68217149,52   
50026                    Não         DESPESAS DE CAPITAL  58132127,48   
50921                    Não         DESPESAS DE CAPITAL  61575657,97   
50932                    Não         DESPESAS DE CAPITAL  18779613,45   
61972                    Não         DESPESAS DE CAPITAL  82323934,88   

       Comprimento  
50025           11  
50026           11  
50921           11  
50932           11  
61972           11

> Observou-se uma inconsistência nos dados, na qual alguns valores apresentam precisão até os centavos, enquanto outros não mantêm a mesma precisão decimal para valores na mesma casa decimal.
- 68217149,52 (tamanho 11)
- 1470678,32 (tamanho 10)
- 872142,22 (tamanho 9)
- 203528,5 (tamanho 8)
- 1358175 (tamanho 7)
- 89035,5 (tamanho 7)

## **Geração de Sentenças**

In [9]:
def generate_sentence(data):
    sentence = (
        f"Em {data['Mês']} de {data['Ano']}, o {data['Tipo Ente']} de {data['Nome Ente']} - {data['UF']}, "
        f"com o código Siafi {data['Código Siafi']} e o código IBGE {data['Código IBGE']}, "
        f"recebeu uma transferência no valor de R$ {data['Valor']}. A {data['Nome Emenda']} "
        f"parlamentar classificada como {data['Categoria Econômica Despesa']} teve como favorecido "
        f"o CNPJ {data['CNPJ do Favorecido']}, identificado pelo nome {data['Nome Favorecido']} "
        f"com a ordem bancária {data['OB']}."
    )
    return sentence

In [10]:
df.head(1)

Nome Ente  UF Código Siafi  Código IBGE   Data   Ano      Mês  \
0  Nova Santa Rosa  PR         7979      4117222  42005  2015  janeiro   

   Tipo Ente                       OB  CNPJ do Favorecido  \
0  Município  135098000012015OB800001      77116663000109   

                Nome Favorecido        Nome Emenda Transferência Especial  \
0  MUNICIPIO DE NOVA SANTA ROSA  Emenda Individual                    Não   

  Categoria Econômica Despesa   Valor  Comprimento  
0         DESPESAS DE CAPITAL  243750            6

"Em **janeiro** de **2015**, o **Município** de **Nova Santa Rosa** - **PR**, com o código Siafi **7979** e o código IBGE **4117222**, recebeu uma transferência no valor de R$ **243750**. A **Emenda Individual** parlamentar classificada como **DESPESAS DE CAPITAL** teve como favorecido o CNPJ **77116663000109**, identificado pelo nome **MUNICIPIO DE NOVA SANTA ROSA** com a ordem bancária **135098000012015OB800001**."


In [11]:
print(generate_sentence(df.iloc[0]))

Em janeiro de 2015, o Município de Nova Santa Rosa - PR, com o código Siafi 7979 e o código IBGE 4117222, recebeu uma transferência no valor de R$ 243750. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 77116663000109, identificado pelo nome MUNICIPIO DE NOVA SANTA ROSA com a ordem bancária 135098000012015OB800001.


In [12]:
sentences = df.apply(generate_sentence, axis=1).to_list()

In [13]:
sentences

['Em janeiro de 2015, o Município de Nova Santa Rosa - PR, com o código Siafi 7979 e o código IBGE 4117222, recebeu uma transferência no valor de R$ 243750. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 77116663000109, identificado pelo nome MUNICIPIO DE NOVA SANTA ROSA com a ordem bancária 135098000012015OB800001.',
 'Em janeiro de 2015, o Município de Humaitá - RS, com o código Siafi 8695 e o código IBGE 4309704, recebeu uma transferência no valor de R$ 144300. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 87613139000199, identificado pelo nome MUNICIPIO DE HUMAITA com a ordem bancária 135098000012015OB800002.',
 'Em janeiro de 2015, o Município de Nova Venécia - ES, com o código Siafi 5677 e o código IBGE 3203908, recebeu uma transferência no valor de R$ 97500. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 27167428000180, identif

## **Tokenização e Geração de Embeddings**

In [14]:
df.shape

(295473, 16)

> O conjunto de dados original possui 295.473 registros, contudo, serão utilizados somente 200 registros para fins de teste dada a limitação de recursos computacionais.

In [15]:
# df_reduced = df[df['Ano'] == 2015].copy()
df_reduced = df.head(200)
df_reduced.shape

(200, 16)

In [16]:
sentences = df_reduced.apply(generate_sentence, axis=1).to_list()

In [17]:
from transformers import AutoTokenizer, AutoModel
import torch

c:\Users\anycs\Documents\MATERIAL_11_PERIODO\PLN\PLN_2024_Silva_Any\venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [18]:
def text_to_embedding(sentences):
    # Load model from HuggingFace Hub
    tokenizer = AutoTokenizer.from_pretrained('rufimelo/Legal-BERTimbau-sts-base-ma-v2')
    model = AutoModel.from_pretrained('rufimelo/Legal-BERTimbau-sts-base-ma-v2')

    # Tokenize sentences
    encoded_input = tokenizer(sentences, padding=True, truncation=True, return_tensors='pt')

    # Compute token embeddings
    with torch.no_grad():
        model_output = model(**encoded_input)
    
    token_embeddings = model_output[0] #First element of model_output contains all token embeddings
    input_mask_expanded = encoded_input['attention_mask'].unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)

In [19]:
sentence_embeddings = text_to_embedding(sentences)

Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


In [20]:
type(sentence_embeddings)

torch.Tensor

In [21]:
sentence_embeddings.size()[0]

200

## **Armazenamento de Embeddings no Milvus**

In [22]:
from pymilvus import connections, utility, MilvusException, \
    FieldSchema, CollectionSchema, DataType, Collection, Index

In [23]:
connections.connect(host="localhost", port="19530")

try:
    collections = utility.list_collections()
    print(f"Lista de coleções:\n", collections)
except MilvusException as e:
    print(e)

Lista de coleções:
 ['amendment_collection']


In [24]:
fields = [
    FieldSchema(name='id', dtype=DataType.INT64, is_primary=True),
    FieldSchema(name='embedding', dtype=DataType.FLOAT_VECTOR, dim=768)
]

In [25]:
schema = CollectionSchema(fields, "Amendments Collection")

In [26]:
collection = Collection('amendment_collection', schema)

In [27]:
index_params = {
    "index_type": "IVF_FLAT",
    "metric_type": "L2", 
    "params" : {"nlist": 100}
}

collection.create_index(field_name="embedding", index_params=index_params)

Status(code=0, message=)

In [31]:
embeddings_list = sentence_embeddings.tolist()

ids = list(range(len(embeddings_list)))


entities = [
    ids,              
    embeddings_list 
]

try:
    collection.insert(entities)
    print("Dados inseridos com sucesso!")
except Exception as e:
    print(f"Erro ao inserir dados: {e}")

Dados inseridos com sucesso!


In [32]:
collection.load()

In [33]:
try:
    # List all collections
    collections = utility.list_collections()
    print(f"List all collections:\n", collections)
except MilvusException as e:
    print(e)

List all collections:
 ['amendment_collection']


## **Busca por Similaridade Semântica**

In [63]:
# Converter a consulta em um embedding
query_text = "2510659"
query_embedding = text_to_embedding(query_text).tolist()
query_embedding

Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.


[[0.17728346586227417,
  -0.03017382137477398,
  -0.3308051526546478,
  -0.21719451248645782,
  0.005756406579166651,
  0.7705256342887878,
  -0.24760626256465912,
  0.28467488288879395,
  0.6963002681732178,
  0.0025434542912989855,
  0.15333248674869537,
  -0.3243366777896881,
  0.12592269480228424,
  -1.2337511777877808,
  -0.3117539584636688,
  -0.11293520778417587,
  -1.2586565017700195,
  0.4336735010147095,
  -0.3780764639377594,
  -0.28897249698638916,
  0.41300952434539795,
  0.01985854096710682,
  0.21573078632354736,
  0.34002554416656494,
  0.5274622440338135,
  -0.08684424310922623,
  -0.5141888856887817,
  -0.15491335093975067,
  0.014990732073783875,
  -0.4190192222595215,
  0.02279781550168991,
  0.1859927773475647,
  -0.1545352190732956,
  -0.49403858184814453,
  -0.002890385687351227,
  0.20740938186645508,
  -0.3341315984725952,
  -0.02704130858182907,
  0.3592727482318878,
  -0.44624149799346924,
  0.11162152141332626,
  0.4609367847442627,
  -0.19240349531173706,
 

In [64]:
# Ajustar parâmetros de pesquisa
search_params = {
    "metric_type": "L2",
    "params": {"nprobe": 10}
}

# Executar a pesquisa
results = collection.search(
    data=query_embedding,  # Use apenas o embedding da consulta
    anns_field="embedding",
    param=search_params,
    limit=10,
    consistency_level="Strong"
)

# Processar resultados
for result in results:
    print(f"IDs: {result.ids}, Distâncias: {result.distances}")

IDs: [154, 155], Distâncias: [101.50675964355469, 102.58380889892578]


In [67]:
sentences[154]

'Em fevereiro de 2015, o Município de São José dos Cordeiros - PB, com o código Siafi 2197 e o código IBGE 2514800, recebeu uma transferência no valor de R$ 125000. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 8873226000109, identificado pelo nome MUNICIPIO DE SAO JOSE DOS CORDEIROS com a ordem bancária 135003000012015OB800208.'

In [68]:
sentences[155]

'Em fevereiro de 2015, o Município de Parari - PB, com o código Siafi 0494 e o código IBGE 2510659, recebeu uma transferência no valor de R$ 125000. A Emenda Individual parlamentar classificada como DESPESAS DE CAPITAL teve como favorecido o CNPJ 1612532000142, identificado pelo nome MUNICIPIO DE PARARI com a ordem bancária 135003000012015OB800209.'